### ISSUE TITLE
Python3.12: type aliases with compound values were not compatiable(was: Problem with `pydantic` model with `Annotated` field with `AfterValidator`

### ISSUE THREAD

**Issue by actionless** [Nuitka/Nuitka#3236]
- Nuitka version, full Python version, flavor, OS, etc. as output by *this exact* command.

  > python -m nuitka --version

```
2.5.4
Commercial: None
Python: 3.12.7 (main, Oct  1 2024, 11:15:50) [GCC 14.2.1 20240910]
Flavor: Arch Python
Executable: ~/test/env/bin/python
OS: Linux
Arch: x86_64
Distribution: Arch None
Version C compiler: /usr/sbin/gcc (gcc 14.2.1).
```

- How did you install Nuitka and Python

`pip`

- The specific PyPI names and versions

  It should be taken from this output if there specific packages involved, e.g. `numpy`, you are
  expected to shorten this to the relevant ones.

  > python -m pip list -v

```
Nuitka                2.5.4
pydantic              2.9.2 
pydantic_core         2.23.4
pydantic-extra-types  2.10.0
pydantic-settings     2.6.1 
pydantic_yaml         1.4.0
```

- Many times when you get an error from Nuitka, your setup may be special

minimal examples work

- Also supply a Short, Self Contained, Correct, Example

```python
import datetime
from typing import Annotated

from pydantic import AfterValidator, AwareDatetime, BaseModel


def some_validation(date: datetime.datetime) -> datetime.datetime:
    return date


type DatetimeUTC = Annotated[AwareDatetime, AfterValidator(some_validation)]  # pylint: disable=invalid-name


class MyDataType(BaseModel):
    date_field: DatetimeUTC | None = None
```

- Provide in your issue the Nuitka options used

- Avoid unnecessary options

```console
$ nuitka print_none.py --output-filename=print-none-bin
$ ./print-none-bin
Traceback (most recent call last):
  File "/home/lie/test/print_none.py", line 14, in <module>
    class MyDataType(BaseModel):
    
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py", line 224, in __new__
    complete_model_class(
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py", line 577, in complete_model_class
    schema = cls.__get_pydantic_core_schema__(cls, handler)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/main.py", line 671, in __get_pydantic_core_schema__
    return handler(source)
           ^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 83, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 655, in generate_schema
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 924, in _generate_schema_inner
    return self._model_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 739, in _model_schema
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1115, in _generate_md_field_schema
    common_field = self._common_field_schema(name, field_info, decorators)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1308, in _common_field_schema
    schema = self._apply_annotations(
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 2107, in _apply_annotations
    schema = get_inner_schema(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 83, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 2088, in inner_handler
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 929, in _generate_schema_inner
    return self.match_type(obj)
           ^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1029, in match_type
    return self._match_generic_type(obj, origin)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1058, in _match_generic_type
    return self._union_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1378, in _union_schema
    choices.append(self.generate_schema(arg))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 650, in generate_schema
    from_property = self._generate_schema_from_property(obj, obj)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 810, in _generate_schema_from_property
    with self.defs.get_schema_or_ref(obj) as (_, maybe_schema):
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/contextlib.py", line 137, in __enter__
    return next(self.gen)
           ^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 2540, in get_schema_or_ref
    ref = get_type_ref(tp)
          ^^^^^^^^^^^^^^^^
  File "/home/lie/test/env/lib/python3.12/site-packages/pydantic/_internal/_core_utils.py", line 93, in get_type_ref
    type_ref = f'{module_name}.{origin.__name__}:{id(origin)}'
                 ^^^^^^^^^^^^^
TypeError: Type NoneType doesn't define __format__
```

**Comment by actionless** [comment]
> Also supply a Short, Self Contained, Correct, Example

offtopic, my OCD's being triggered by this asymmetry, it must be instead:

`Short, Self Contained, Correct, Eligible Example`

**Comment by Rogdham** [comment]
**Workaround**

Remove the `type` keywordseems to prevent the issue (in original message, remove `type` in `type DatetimeUTC = ...`).

---

Reproduced without Pydantic:
```python
from typing import Annotated

def foo(origin):
    module_name = getattr(origin, '__module__', '<No __module__>')
    type_ref = f'{module_name}.{origin.__name__}:{id(origin)}'
    return type_ref

type Bar = Annotated[int, 42]

print(foo(Bar))
```

```
Traceback (most recent call last):
  File "/redacted/a.py", line 10, in <module>
    print(foo(Bar))
    
  File "/redacted/a.py", line 5, in foo
    type_ref = f'{module_name}.{origin.__name__}:{id(origin)}'

TypeError: Type NoneType doesn't define __format__
```

---

It seems that trying to access the `__module__` attribute on a type variable defined with PEP-695 syntax results in an error:
```py
from typing import Annotated

type Bar = Annotated[int, 42]

print(hasattr(Bar, "__module__"))
```
```
Traceback (most recent call last):
  File "/redacted/b.py", line 5, in <module>
    print(hasattr(Bar, "__module__"))
    
SystemError: Objects/funcobject.c:273: bad argument to internal function
```

**Comment by kayhayen** [comment]
> Short, Self Contained, Correct, Eligible Example

I am taking that one, going to be on develop soon.

**Comment by kayhayen** [comment]
Not sure now, if I only forgot about this one, I remember seeing it, the original code doesn't crash with 2.6 anymore. I am checking the other examples that were given off-topic (?) here too.

**Comment by kayhayen** [comment]
Somehow, for the type alias, we do not set the module value in the struct, no idea how that came to pass. Somehow, I think that at least the "value" in empty, from my recollection.

**Comment by kayhayen** [comment]
For the module value, I have a fix on the factory branch, but the original code example still doesn't work, I need to look at why value isn't populated either. I believe I found that somehow these were unused in Python 3.12 for regular declarations, but I was very wrong obviously. I will also put the change in 2.6.3 just because I can, and crashing is never a good thing, but the issue shall remain open to be worked on.

**Comment by Rogdham** [comment]
I can confirm that my 2 reproducible cases (accessing `__module__`) are fixed with 2.6.3. 🎉 

<details><summary>Here is the error in 2.6.3 with the original Pydantic case.</summary>
<p>

```
Traceback (most recent call last):
  File "/redacted/run.py", line 14, in <module>
    class MyDataType(BaseModel):
    
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py", line 226, in __new__
    complete_model_class(
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_model_construction.py", line 658, in complete_model_class
    schema = cls.__get_pydantic_core_schema__(cls, handler)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/main.py", line 702, in __get_pydantic_core_schema__
    return handler(source)
           ^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 84, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 612, in generate_schema
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 881, in _generate_schema_inner
    return self._model_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 693, in _model_schema
    {k: self._generate_md_field_schema(k, v, decorators) for k, v in fields.items()},
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1073, in _generate_md_field_schema
    common_field = self._common_field_schema(name, field_info, decorators)
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1265, in _common_field_schema
    schema = self._apply_annotations(
             ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 2062, in _apply_annotations
    schema = get_inner_schema(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 84, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 2043, in inner_handler
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 886, in _generate_schema_inner
    return self.match_type(obj)
           ^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 988, in match_type
    return self._match_generic_type(obj, origin)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1016, in _match_generic_type
    return self._union_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1327, in _union_schema
    choices.append(self.generate_schema(arg))
                   ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 612, in generate_schema
    schema = self._generate_schema_inner(obj)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 886, in _generate_schema_inner
    return self.match_type(obj)
           ^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 950, in match_type
    return self._type_alias_type_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 1359, in _type_alias_type_schema
    schema = self.generate_schema(annotation)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 607, in generate_schema
    from_property = self._generate_schema_from_property(obj, obj)
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 761, in _generate_schema_from_property
    schema = get_schema(
             ^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/types.py", line 2249, in __get_pydantic_core_schema__
    schema = handler(source)
             ^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_schema_generation_shared.py", line 84, in __call__
    schema = self._handler(source_type)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 886, in _generate_schema_inner
    return self.match_type(obj)
           ^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 997, in match_type
    return self._unknown_type_schema(obj)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/redacted/.venv/lib/python3.12/site-packages/pydantic/_internal/_generate_schema.py", line 515, in _unknown_type_schema
    raise PydanticSchemaGenerationError(
pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for AwareDatetime. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.

If you got this error by calling handler(<some type>) within `__get_pydantic_core_schema__` then you likely need to call `handler.generate_schema(<some type>)` since we do not call `__get_pydantic_core_schema__` on `<some type>` otherwise to avoid infinite recursion.

For further information visit https://errors.pydantic.dev/2.10/u/schema-for-unknown-type
```


</p>
</details>

**Comment by kayhayen** [comment]
Yeah, I meant to report here, that the module part is fixed. I will have to research what `value` is vs `computed value` in Python there, and it probably will resolve all issues, but this part was obviously very trivial to fix in comparison and may help a number of other codes already. I will continue this for a later hotfix.

**Comment by kayhayen** [comment]
On the C side, this is provided:

```
MAKE_TYPE_ALIAS:
()
typing.Annotated[pydantic.types.AwareDatetime, AfterValidator(func=<compiled_function some_validation at 0x7fb692552a40>)]
```

So, looks good to me, probably a matter of confusing computed value and value in the structure.

**Comment by kayhayen** [comment]
One score up in the eternal battle of experimental computer science vs. actual knowledge, switching the two around, makes it work with the original example.

**Comment by kayhayen** [comment]
So, this is indeed now fully fixed on factory, the original code example works, also in standalone mode, but that was to be expected and it will come to the next hotfix as well.

**Comment by actionless** [comment]
is there some specific guide how `factory` version could be installed, or i should just do `pip install <git_repo>#branch=factory`?

also if i recall correctly it was still happening after python updated to 3.13 in arch linux (but i'll need to double check when back at pc on which i develop that project)

**Comment by kayhayen** [comment]
I thought I had posted this for the first part of the fix:

Thanks for your report. This is fixed on the factory branch, which is a development version that is under rapid development. You can try it out by going here: https://nuitka.net/doc/factory.html 
	
Feedback on whether this is working is very welcome. Please do not share plans to do it; only confirm or deny that it is working.

**Comment by Rogdham** [comment]
Hello, on my side, all reproducible tests I had on this issue, as well as real code, are now properly working when using the factory branch.

Thank you for the fix!

**Comment by kayhayen** [comment]
Part of the 2.6.6 hotfix I just made.
